Fluentd with fluentbit

nutellinoit opened the pull request:
Hi guys, i'm opening this PR as a starting point to solve the json parsing problem.

Seeing what banzaicloud is doing with their operator, i studied the logging structure and using this architecture solves the json parsing (fluent-bit is very easy to setup)

![flow-chart](https://user-images.githubusercontent.com/25492971/82608348-c637f600-9bba-11ea-8090-3846ae2358ff.png)

What do you think?

phisco commented:
I’ll give a more technical feedback tomorrow morning, but in the meantime: YES! YES! YES! ❤️

lnovara commented:
Tomorrow morning I will do an in-depth review, so far it looks good.

This is the logging architecture we were talking about some time ago, right @phisco?

phisco commented:
Exactly @lnovara! The same architecture adopted by Anthos

phisco commented on katalog/fluentd/deploy.yml:
```diff
@@ -1,10 +1,11 @@
 apiVersion: apps/v1
-kind: DaemonSet
+kind: Deployment
```
Any reason for adopting a Deployment instead of a Statefulset?
I've seen fluentd in this configuration usually deployed as the latter to allow persistent buffering from fluentd.

nutellinoit replied:
As soon as i understand how buffering works, i will change it as statefulset

phisco replied:
in the eleasticsearch output configuration there is the buffer section (https://docs.fluentd.org/configuration/buffer-section), as you can see currently we are using in memory buffer, but we could use files and move to statefulset to be sure no logs are lost after restarts.

nutellinoit replied:
done!

phisco commented on katalog/fluentd/README.md:
```diff
@@ -1,9 +1,11 @@
 collected data in various destinations. Fluentd deployment in Fury Kubernetes
 Logging is configured to write records into Elasticsearch.
 
+Fluent-bit is an open source log collector and forwarder. It sends all collected logs to fluentd. TODO
```
what's to be done?

nutellinoit replied:
I didn't finish writing the README, it's a placeholder 😊

angelbarrera92 commented:
In the architecture shown in the PR description, fluentd lives outside the cluster.
Just to keep in mind. Modify the diagram if you plan to include it in the docs or clarify our use case.

BTW: Awesome work.

One more question.
Are all the logs parsed as it was with fluentbit? including those related with audit?

Thanks

nutellinoit commented:
> In the architecture shown in the PR description, fluentd lives outside the cluster.
> Just to keep in mind. Modify the diagram if you plan to include it in the docs or clarify our use case.

I can modify the image, i copy pasted it from the fluentd website

> One more question.
> Are all the logs parsed as it was with fluentbit? including those related with audit?

I need to check this

nutellinoit commented:
Ok, i added some other systemd on fluentbit and thanks to @lzecca78 systemd logging gathering is working! (mounting machine-id juju)

I think I will also need to add the auto elasticsearch index templating (like @lnovara told me) .

In the end, a little **breaking change**:

Moved all workloads logs to: `containers-*` and all the system logs (audit, systemd, etc): `system-*` indexes. I created also new actions on curator to rotate new indexes.

This change was due to too many fields on a single index, and to separate "system logs" from "workloads log".

nutellinoit commented:
Quick poll on the workload index name:

* 👍  `containers-*`
* 🚀  `workloads-*`
* ☁️  `kubernetes-*`

lzecca78 commented:
could make sense even to add the curator section that is in charge to archive index older than X to an object storage (for this is needed a plugin in elasticsearch for it)?

phisco commented on katalog/fluentd/fluent.conf:
```diff
@@ -71,11 +70,10 @@
      queue_limit_length "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_QUEUE_LIMIT_LENGTH'] || '32'}"
      retry_max_interval "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_RETRY_MAX_INTERVAL'] || '30'}"
      retry_forever true
+     overflow_action drop_oldest_chunk
```
Any reason for not setting block? (doc [here](https://docs.fluentd.org/configuration/buffer-section), example  [here](https://github.com/helm/charts/blob/master/stable/fluentd/values.yaml#L149)

nutellinoit replied:
I thought "drop_oldest_chunk" could help when system is overflowed with logs, but if "block" helps in the same way i can change to it.

I think fluentbit will complain about being unable to send logs to fluentd right?

phisco replied:
block should backpressure to the input plugin instead of dropping logs, not sure how it is propagated to fluentbit

phisco replied:
https://fluentbit.io/documentation/0.13/configuration/backpressure.html here the doc

nutellinoit replied:
Changed to block, and in fluent-bit there is infinite retry, so i think it's ok!

phisco commented on katalog/fluentd/fluent.conf:
```diff
@@ -114,5 +112,6 @@
      queue_limit_length "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_QUEUE_LIMIT_LENGTH'] || '32'}"
      retry_max_interval "#{ENV['FLUENT_ELASTICSEARCH_BUFFER_RETRY_MAX_INTERVAL'] || '30'}"
      retry_forever true
+     overflow_action drop_oldest_chunk
```
same here, why not block?

phisco commented:
> Quick poll on the workload index name:
> 
>     * 👍  `containers-*`
> 
>     * 🚀  `workloads-*`
> 
>     * ☁️  `kubernetes-*`

☁️ is not an option for quick reactions to comments! KOMPLOTTO!!!!111!!!

(I'm voting for ☁️)

phisco commented on katalog/fluentd/fluent-bit.conf:
```diff
@@ -0,0 +1,62 @@
+[SERVICE]
+    Flush        1
+    Daemon       Off
+    Log_Level    info
+    Parsers_File parsers.conf
+    storage.path  /buffers
```
we could add the prometheus exporter and a servicemonitor https://docs.fluentbit.io/manual/administration/monitoring

nutellinoit replied:
@phisco done!

lnovara commented:
> Quick poll on the workload index name:
> 
>     * 👍  `containers-*`
> 
>     * 🚀  `workloads-*`
> 
>     * ☁️  `kubernetes-*`
☁️

nutellinoit commented:
☁️ too

lzecca78 commented:
☁️  also for me.

nutellinoit commented:
`kubernetes-*` sold!

angelbarrera92 commented on katalog/fluentd/fluentbit.yaml:
```diff
@@ -0,0 +1,99 @@
+      tolerations:
+        - operator: Exists
+      containers:
+        - image: fluent/fluent-bit:1.4.4
```
Try yo put the image tag in the `kustomization.yaml` file please ;)

nutellinoit replied:
Will do!

angelbarrera92 commented:
Looks really cool.

But i think we lose the nginx parser we implemented in latest release. Isn't it?
Also.... How is the update path?

Thanks!

nutellinoit commented:
> Looks really cool.
> 
> But i think we lose the nginx parser we implemented in latest release. Isn't it?
> Also.... How is the update path?
> 
> Thanks!

I can re-implement it

nutellinoit commented:
@angelbarrera92 added a "draft" release note, re-added ingress-controller-*

angelbarrera92 commented:
> @angelbarrera92 added a "draft" release note, re-added ingress-controller-*

Did you checked that is parsing correctly? (nginx)

nutellinoit commented:
> > @angelbarrera92 added a "draft" release note, re-added ingress-controller-*
> 
> Did you checked that is parsing correctly? (nginx)

Yes it is, i'm doing some more tests, also trying to add a "deadletter" index for all the stuff rejected from ES if there are index mapping problems

nutellinoit commented:
Ok, i did some tests:

* on the **system** log, enabling 400 error output on fluentd shows that a lot of audit logs are discarded silently due to "mapping conflict" or "exceeding 1000 fields". This way we are not saving all audit stuff
  * my idea is to change the INPUT on fluentbit removing the json parse for the audit logs and 
    also adding another index pattern only for audit : `audit-*` , segregating all types of logs in 
    different indexes
* also, if there are problems on elasticsearch, like congestion, fluentd re-sends the same records without setting the `_id` on elasticsearch document. This cause a lot of duplicated records
  * this can be fixed adding this lines on the configuration:
    ```
    <filter kubernetes.**>
       @type elasticsearch_genid
       hash_id_key _hash # storing generated hash id key (default is _hash)
    </filter>
    <match kubernetes.**>
      @type elasticsearch
      @id out_es_kubernetes
      @log_level "#{ENV['FLUENT_LOG_LEVEL'] || 'error'}"
      id_key _hash # specify same key name which is specified in hash_id_key
      remove_keys _hash # Elasticsearch doesn't like keys that start with _
      .....
    ```

What do you think?

angelbarrera92 commented:
> Ok, i did some tests:
> 
>     * on the **system** log, enabling 400 error output on fluentd shows that a lot of audit logs are discarded silently due to "mapping conflict" or "exceeding 1000 fields". This way we are not saving all audit stuff
>       
>       * my idea is to change the INPUT on fluentbit removing the json parse for the audit logs and
>         also adding another index pattern only for audit : `audit-*` , segregating all types of logs in
>         different indexes


Do it ;) 

>     * also, if there are problems on elasticsearch, like congestion, fluentd re-sends the same records without setting the `_id` on elasticsearch document. This cause a lot of duplicated records
>       
>       * this can be fixed adding this lines on the configuration:
>         ```
>         <filter kubernetes.**>
>            @type elasticsearch_genid
>            hash_id_key _hash # storing generated hash id key (default is _hash)
>         </filter>
>         <match kubernetes.**>
>           @type elasticsearch
>           @id out_es_kubernetes
>           @log_level "#{ENV['FLUENT_LOG_LEVEL'] || 'error'}"
>           id_key _hash # specify same key name which is specified in hash_id_key
>           remove_keys _hash # Elasticsearch doesn't like keys that start with _
>           .....
>         ```
> 
> 
> What do you think?

Should works!

nutellinoit commented:
Also, i would like to change the default to FLUENT_ELASTICSEARCH_LOG_ES_400_REASON from `false` to `true`, this way we can "save" the problematic logs using fluentd stdout as medium

nutellinoit commented:
@lnovara added automatic index template creation on ES

lnovara commented on katalog/fluentd/fluentd-index-sighup-template.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "index_patterns" : ["system-*","kubernetes-*","ingress-controller-*","audit-*"],
+  "settings": {
+    "number_of_shards": 3,
```
I would use `number_of_shards: 1` and `number_of_replicas: 2`, our use case wouldn't benefit much from the sharding.

nutellinoit replied:
Ok, changing to 1 shard and 2 replicas

lnovara commented on katalog/fluentd/kustomization.yaml:
```diff
@@ -1,22 +1,32 @@
+    - kubernetes-output.conf=conf.d/kubernetes-output.conf
+    - system-output.conf=conf.d/system-output.conf
+    - ingress-controller-output.conf=conf.d/ingress-controller-output.conf
+    - audit-output.conf=conf.d/autit-output.conf
```
Typo: `autit`

nutellinoit replied:
Will fix!

lnovara commented on katalog/fluentd/prometheus.conf:
```diff
@@ -1,11 +0,0 @@
-# Prometheus metric exposed on 0.0.0.0:24231/metrics
```
Why are we dropping Prometheus metrics from fluentd?

nutellinoit replied:
This setting is on the main fluentd.conf file

lnovara replied:
Sorry, I totally missed it.

lnovara commented:
Should we add a note about how to properly patch the index template in case someone wants/needs to use `elasticsearch-single`?

nutellinoit commented:
> Should we add a note about how to properly patch the index template in case someone wants/needs to use `elasticsearch-single`? Otherwise we end up with unassigned replicas.

I added on the draft release note how to patch the template, i can add a note about elasticsearch-single

nutellinoit commented:
Updates:

* Changed log_es_400_reason from false to true by default
* Changed template to 1 shard and 2 replicas by default
* Added on the release note how to patch fluentd if you are using `elasticsearch-single`

phisco commented:
LGTM

lnovara commented on docs/releases/draft.md:
````diff
@@ -0,0 +1,78 @@
+```yaml
+configMapGenerator:
+  - name: fluentd-index-template
+    behavior: replace
````
Please, add `namespace: logging` to the `configMapGenerator` otherwise it may not work as expected.

nutellinoit replied:
Done!

lnovara commented on docs/releases/draft.md:
````diff
@@ -0,0 +1,78 @@
+
+```yaml
+configMapGenerator:
+  - name: fluentd-index-template
````
Missing `namespace`.

nutellinoit replied:
Done!

angelbarrera92 commented on docs/releases/draft.md:
```diff
@@ -0,0 +1,80 @@
+
+## Changelog
+
+Changes between `1.3.0` and this release: `X.X.X`
```
Just change X.X.X -> 1.4.0 and i think it is ready to be merged!

Am i missing something?

lzecca78 commented:
![](https://gifimage.net/wp-content/uploads/2017/09/92-minuti-di-applausi-gif-1.gif)

nutellinoit commented:
@angelbarrera92 ready to merge!
